contracts/src/v0.8/dev/automation/2_1/KeeperRegistry2_1.sol

```diff
@@ -461,42 +449,52 @@ contract KeeperRegistry2_1 is KeeperRegistryBase2_1, OCR2Abstract, Chainable, ER
    */
   function _validateBlockTrigger(
     uint256 upkeepId,
-    BlockTrigger memory trigger,
+    bytes memory rawTrigger,
     Upkeep memory upkeep
   ) internal returns (bool) {
-    if (trigger.blockNum < upkeep.lastPerformed) {
+    BlockTrigger memory trigger = abi.decode(rawTrigger, (BlockTrigger));
+    if (trigger.blockNum < upkeep.lastPerformedBlockNumberOrTimestamp) {
       // Can happen when another report performed this upkeep after this report was generated
-      emit StaleUpkeepReport(upkeepId);
+      emit StaleUpkeepReport(upkeepId, rawTrigger);
       return false;
     }
     if (_blockHash(trigger.blockNum) != trigger.blockHash) {
```

Comment:
currently this offers protection against 'future' block numbers as well as standard re-org protection.
Since we want to support skipping reorg protection by setting blockHash=0, this logic needs more thinking if the two checks need to be separated